libparsec/crates/types/src/fs_path.rs

```diff
@@ -248,6 +248,22 @@ impl FsPath {
     pub fn from_parts(parts: Vec<EntryName>) -> Self {
         Self { parts }
     }
+
+    /// Normalize the path
+    pub fn normalize(&self) -> Self {
+        let mut new_parts = Vec::with_capacity(self.parts.len());
+
+        for part in &self.parts {
+            match part.0.as_str() {
+                "." | "" => continue,
+                ".." => {
+                    new_parts.pop();
+                }
+                _ => new_parts.push(part.clone()),
+            }
+        }
+        Self { parts: new_parts }
+    }
```

Comment:
Maybe we don't need to allocate a new `path`, it is better if we have the normalization. WDYT ?

```suggestion
    /// Normalize the path
    pub fn normalize(self) -> Self {
        let mut new_parts = Vec::with_capacity(self.parts.len());

        for part in &self.parts {
            match part.0.as_str() {
                "." | "" => continue,
                ".." => {
                    new_parts.pop();
                }
                _ => new_parts.push(part),
            }
        }
        Self { parts: new_parts }
    }
```

Reply:
I've taken inspiration of [`Path::canonicalize`](https://doc.rust-lang.org/std/path/struct.Path.html#method.canonicalize) where it take a ref.

With your suggestion I would need to clone `FsPath` before using it :shrug:

Reply:
cloning the `vec` is faster than cloning each element `1` by `1`, and why do you want to keep the old one ? In my opinion, most of the time you would like to drop the older one :thinking:

Reply:
You never know :shrug:,

I've applied your suggestion yesterday.